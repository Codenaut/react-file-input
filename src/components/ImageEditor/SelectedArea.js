import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import '../../styles/ImageEditor.scss';

class SelectedArea extends PureComponent {
  constructor(props) {
    super(props);

    this.selectedArea = null;

    this.startMove = this.startMove.bind(this);

    this.resizeTopLeft = this.resizeTopLeft.bind(this);
    this.resizeTop = this.resizeTop.bind(this);
    this.resizeTopRight = this.resizeTopRight.bind(this);
    this.resizeRight = this.resizeRight.bind(this);
    this.resizeBottomRight = this.resizeBottomRight.bind(this);
    this.resizeBottom = this.resizeBottom.bind(this);
    this.resizeBottomLeft = this.resizeBottomLeft.bind(this);
    this.resizeLeft = this.resizeLeft.bind(this);
  }

  resizeTopLeft(event) {
    const { startResize } = this.props;

    startResize(event, true, true, false);
  }

  resizeTop(event) {
    const { startResize } = this.props;

    startResize(event, false, true, true, true);
  }

  resizeTopRight(event) {
    const { startResize } = this.props;

    startResize(event, false, true);
  }

  resizeRight(event) {
    const { startResize } = this.props;

    startResize(event, false, false, true);
  }

  resizeBottomRight(event) {
    const { startResize } = this.props;

    startResize(event, false, false);
  }

  resizeBottom(event) {
    const { startResize } = this.props;

    startResize(event, false, false, true, true);
  }

  resizeBottomLeft(event) {
    const { startResize } = this.props;

    startResize(event, true, false);
  }

  resizeLeft(event) {
    const { startResize } = this.props;

    startResize(event, true, false, true);
  }

  startMove(event) {
    if (event.target === this.selectedArea) {
      const { startMove } = this.props;

      startMove(event);
    }
  }

  render() {
    const { style } = this.props;

    return (
      <div
        className="brainhub-image-editor__selected-area"
        ref={ref => {
          this.selectedArea = ref;
        }}
        style={style}
        onMouseDown={this.startMove} >
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--tl"
          onMouseDown={this.resizeTopLeft}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--t"
          onMouseDown={this.resizeTop}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--tr"
          onMouseDown={this.resizeTopRight}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--r"
          onMouseDown={this.resizeRight}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--br"
          onMouseDown={this.resizeBottomRight}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--b"
          onMouseDown={this.resizeBottom}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--bl"
          onMouseDown={this.resizeBottomLeft}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--l"
          onMouseDown={this.resizeLeft}
        />
      </div>
    );
  }
}

SelectedArea.defaultProps = {
  style: {
    top: '0',
    right: '0',
    bottom: '0',
    left: '0',
  },
};

SelectedArea.propTypes = {
  style: PropTypes.shape({
    top: PropTypes.string,
    right: PropTypes.string,
    bottom: PropTypes.string,
    left: PropTypes.string,
  }),
  startMove: PropTypes.func.isRequired,
  startResize: PropTypes.func.isRequired,
};

export default SelectedArea;
