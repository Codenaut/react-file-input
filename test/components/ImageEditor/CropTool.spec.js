import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import CropTool from '../../../src/components/ImageEditor/CropTool';
import SelectedArea from '../../../src/components/ImageEditor/SelectedArea';

const defaultProps = {
  image: '',
  onCancelEdition: () => null,
  onSaveEdition: () => null,
};

const setup = (props = {}) => {
  const cropTool = mount(<CropTool {...defaultProps} {...props} />);

  return {
    cropTool,
    selectedArea: cropTool.find(SelectedArea),
    button: cropTool.find('button'),
  };
};

describe('components', () => {
  describe('ImageEditor', () => {
    describe('CropTool', () => {
      it('should render with a default state', () => {
        const { cropTool } = setup();
        expect(cropTool.state()).toEqual({
          landscape: true,
          isMoving: false,
          isResizing: false,
          resizeVertical: false,
          x0: 0,
          y0: 0,
          x1: 0,
          y1: 0,
          xM: 0,
          yM: 0,
        });
      });

      it('should render a SelectedArea', () => {
        const { selectedArea } = setup();

        expect(selectedArea).toHaveLength(1);
      });

      it('should render two buttons', () => {
        const { button } = setup();

        expect(button).toHaveLength(2);
        expect(button.first().text()).toBe('Cancel');
        expect(button.last().text()).toBe('Save image');
      });

      it('should cancel render when Cancel button is clicked', () => {
        const onCancelEdition = jest.fn();
        const { button } = setup({ onCancelEdition });

        const cancelButton = button.first();

        cancelButton.simulate('click');

        expect(onCancelEdition).toHaveBeenCalled();
      });

      it('should save the edition render when Save button is clicked', () => {
        const onSaveEdition = jest.fn();
        const { cropTool, button } = setup({ onSaveEdition });

        const saveButton = button.last();

        saveButton.simulate('click');

        const { x0, y0, x1, y1 } = cropTool.state();
        expect(onSaveEdition).toHaveBeenCalledWith(cropTool.instance().image, { x0, y0, x1, y1 });
      });

      it('should start moving when startMove is called', () => {
        const { cropTool } = setup();
        const event = {
          preventDefault: () => null,
        };
        cropTool.instance().startMove(event);
        expect(cropTool.state('isMoving')).toBeTruthy();
      });

      it('should only interact to mouse movement is the state isMoving or isResizing', () => {
        const { cropTool } = setup();
        const event = {
          preventDefault: () => null,
        };
        const cropToolInstance = cropTool.instance();
        const handleMoveSpy = jest.spyOn(cropToolInstance, 'handleMove');
        const handleReiszeSpy = jest.spyOn(cropToolInstance, 'handleResize');

        cropToolInstance.onMouseMove(event);

        expect(handleMoveSpy).not.toHaveBeenCalled();
        expect(handleReiszeSpy).not.toHaveBeenCalled();

        cropTool.setState({ isMoving: true });
        cropToolInstance.onMouseMove(event);

        expect(handleMoveSpy).toHaveBeenCalledTimes(1);
        expect(handleReiszeSpy).not.toHaveBeenCalled();

        cropTool.setState({ isMoving: false, isResizing: true });
        cropToolInstance.onMouseMove(event);

        expect(handleMoveSpy).toHaveBeenCalledTimes(1);
        expect(handleReiszeSpy).toHaveBeenCalledTimes(1);
      });

      it('should unset isMoving and isResizing and order the selected area points when mouseup', () => {
        const { cropTool } = setup();

        const prevState = {
          landscape: true,
          isMoving: false,
          isResizing: true,
          isResizingSide: true,
          resizeVertical: true,
          x0: 50,
          y0: 50,
          x1: 10,
          y1: 10,
          xM: 10,
          yM: 10,
        };
        cropTool.setState(prevState);

        const event = {
          preventDefault: () => null,
        };

        cropTool.instance().onMouseUp(event);

        expect(cropTool.state()).toEqual({
          ...prevState,
          isMoving: false,
          isResizing: false,
          isResizingSide: false,
          resizeVertical: false,
          x0: 10,
          y0: 10,
          x1: 50,
          y1: 50,
          xM: 0,
          yM: 0,
        });
      });

      it('should match exact snapshot', () => {
        const tree = renderer.create(
          <div>
            <CropTool {...defaultProps} />
          </div>
        ).toJSON();

        expect(tree).toMatchSnapshot();
      });
    });
  });
});
