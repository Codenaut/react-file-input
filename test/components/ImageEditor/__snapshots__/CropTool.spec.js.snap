// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components ImageEditor CropTool should match exact snapshot 1`] = `
<div>
  <div>
    <div
      className="brainhub-crop-tool"
      onDragStart={[Function]}
    >
      <img
        className="brainhub-crop-tool__image"
        onLoad={[Function]}
        src=""
      />
      <div
        className="brainhub-selected-area"
        onMouseDown={[Function]}
        style={Object {}}
      >
        <div
          className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--tl"
          onMouseDown={[Function]}
        />
        <div
          className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--t"
          onMouseDown={[Function]}
        />
        <div
          className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--tr"
          onMouseDown={[Function]}
        />
        <div
          className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--r"
          onMouseDown={[Function]}
        />
        <div
          className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--br"
          onMouseDown={[Function]}
        />
        <div
          className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--b"
          onMouseDown={[Function]}
        />
        <div
          className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--bl"
          onMouseDown={[Function]}
        />
        <div
          className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--l"
          onMouseDown={[Function]}
        />
      </div>
    </div>
    <div
      className="brainhub-crop-tool__controls"
    >
      <button
        className="brainhub-crop-tool__button brainhub-crop-tool__button--cancel"
        onClick={[Function]}
      >
        Cancel
      </button>
      <button
        className="brainhub-crop-tool__button brainhub-crop-tool__button--save"
        onClick={[Function]}
      >
        Save image
      </button>
    </div>
  </div>
</div>
`;
